User class reformatting, restructuring, renaming, and debugging modifications

bahamieh opened the pull request:
### Overview 
This pull request helps make clear distinctions between different function variables, create a cohesive format of javadocs and print statements, gets rid of unneeded comments, fixes a bug, to increase consistency, usability and readability. 

Modifications to the User class include:
* **Rewriting of the user class java doc**
 * Illustrates a more detailed view of what the user class methods accomplish
* **Inclusion of imports of common java.lang keywords**
  * Avoids repeatedly using `System.` in front of each keyword every time they are needed
* **Reformatting and rewriting of print statements**
  * Concatenates some of the print statements so that they take less lines and less print statements while achieving the same output
  * Expands other print statements so that all print statements have the same structure and readability for both the users and readers.
  * Rewrites some of the print statements for a more user friendly experience.
* **Removing of unnecessary and insignificant comments**
* **Rewriting and reformatting of javadocs for each function**
  * Reformats so that all functions which do the same tasks for different variable inputs have the same javadoc structure
  * Places periods at the end of all full sentence comments per the Java Style Guide examples
* **Renaming of variables in verification functions for uniqueness**
  * Reformats so that all functions which do the same tasks for different variable inputs have the same javadoc structure
* **Renaming of variables in verification functions for uniqueness**
  * Renames the variables which are passed into all the verification functions so that they are unique to their purpose
    * Instead of having toVerify as the variable for both the username and host name functions, they are now `usernameToVerify` and `hostnameToVerify`. The password variable has been changed to `passwordToVerify` as well
* **Renaming of pattern variables specific to their input type**
  * Renames both the username pattern and hostname pattern variables which were named userNamePattern. This is confusing so two separate variables were made to distinguish the two
   * The capital case for Name in userNamePattern is not consistent with the rest of the code which keeps it lowercase so that was changed to `usernamePattern` instead of `userNamePattern`
* **Fixing of username input bug**
   * Fixes a newly discovered username input verification bug which only has the user re-enter a username if invalid once and after than would accept an invalid username and move on to the password
   * The program now will ask to re-enter a username as many times as needed until the username is valid before moving on to entering the password
* **Fixing of inconsistency in error handling output**
   * Originally prints out as a regular print statement instead of it being an error print statement just like the username and password methods
    *This has now been changed so that all error handling uses `err.println` for consistency and clarity when reading and using the program

bahamieh committed:
Rewrite user class java doc

Rewrote the user class java doc to have more details on what the user class methods accomplish.

bahamieh committed:
Add imports of common java.lang keywords

Added imports of common java.lang keywords "out", "err", and "in" to avoid repeatedly using "System." infront of each keyword everytime they are needed in order to improve readability.

bahamieh committed:
Reformat and rewrote print statements

Reformatted and concatenated some of the print statements so that they take less lines and less print statements while achieving the same output. This was done for better clarity and structure.

Expanded the others so that all print statements have the same structure and readability for both the users and readers.

Rewrote some of the print statements for a more user friendly experience.

The above actions were done so that the code looks cohesive, is easier to follow and read, and has a rhythm and pattern throughout for better detections of code segments.

bahamieh committed:
Remove unnecessary comments

Removed two comments that are insignificant.

bahamieh committed:
Rewrite and reformat javadoc of each function

Rewrote and reformated javadocs so that all functions which do the same tasks for different variable inputs have the same javadoc structure. This was done so that is it is easier to recognize how functions relate and patterns in the code. Also added periods at the end of all full sentence comments per the Java Style Guide examples.

bahamieh committed:
Rename variables in verification functions for uniqueness

Renamed the variables which are passed into all the verification functions so that they are unique to their purpose. Instead of having toVerify as the variable for both the username and host name functions, they are now usernameToVerify and hostnameToVerify. The password variable has been changed to passwordToVerify as well. This is done to help make a clear distinctions between different function variables and to reduce confusion.

bahamieh committed:
Rename pattern variables specific to their input type

Renamed the pattern strings so that they are specific to their input type. Both the username pattern and hostname pattern variables were named userNamePattern which is confusing so two separate variables were made to distinguish the two. The capital case for Name in userNamePattern is not consistent with the rest of the code which keeps it lowercase so that was changed to "usernamePattern" instead of "userNamePattern".

bahamieh committed:
Fix username input bug

Fixed a newly discovered username input verification bug which only has the user re-enter a username if invalid once and after than would accept an invalid username and move on to the password. The program now will ask to re-enter a username as many times as needed until the username is valid before moving on to entering the password.

bahamieh committed:
Fix inconsistency in error handling output

The host name error handling output used to print out as a regular print statement instead of it being an error print statement just like the username and password methods. This has now been changed so that all error handling uses err.println for consistency and clarity when reading and using the program.

carissaallen commented on src/main/java/edu/pdx/cs/sftp/User.java:
```diff
@@ -44,93 +51,97 @@
 
   /**
-   * Prompt the user for a valid password
+   * Prompt the user for a valid password.
```
You might consider making sure all the verb forms end in `-s` throughout the javadoc in this PR since you're making changes to the class as a whole. 
```suggestion
   * Prompts the user for a valid password.
```

rooneyshuman replied:
```suggestion
   * Prompts the user for a valid password.
```

carissaallen commented on src/main/java/edu/pdx/cs/sftp/User.java:
```diff
@@ -44,93 +51,97 @@
 
   /**
-   * Verify the string isn't empty and doesn't contain spaces
+   * Verify the password isn't empty and doesn't contain spaces.
```
```suggestion
   * Verifies the password isn't empty and doesn't contain spaces.
```

rooneyshuman replied:
```suggestion
   * Verifies the password isn't empty and doesn't contain spaces.
```

carissaallen commented on src/main/java/edu/pdx/cs/sftp/User.java:
```diff
@@ -44,93 +51,97 @@
 
   /**
-   * Prompts the user for a valid username (usernames must be alphanumeric of size 8-20)
+   * Prompt the user for a valid username.
```
```suggestion
   * Prompts the user for a valid username.
```

rooneyshuman replied:
```suggestion
   * Prompts the user for a valid username.
```

carissaallen commented on src/main/java/edu/pdx/cs/sftp/User.java:
```diff
@@ -44,93 +51,97 @@
 
   /**
-   * Checks if the string is a valid username
+   * Verify the username is 8-20 alphanumeric characters.
```
```suggestion
   * Verifies the username is 8-20 alphanumeric characters.
```

rooneyshuman replied:
```suggestion
   * Verifies the username is 8-20 alphanumeric characters.
```

carissaallen commented on src/main/java/edu/pdx/cs/sftp/User.java:
```diff
@@ -44,93 +51,97 @@
    *
-   * @param toVerify a string to be checked against a regular expression
-   * @return <code>true</code> if the string is a valid username <code>false</code> otherwise
+   * @param usernameToVerify is the string to be checked against a regular expression.
```
This parameter name change is much clearer! 👍

rooneyshuman replied:
```suggestion
   * @param usernameToVerify the input to be checked against a regular expression.
```

carissaallen commented on src/main/java/edu/pdx/cs/sftp/User.java:
```diff
@@ -44,93 +51,97 @@
-   * Prompts the user for a valid host name which follows the following parameters: (1) Must be
-   * alphanumeric + "." (2) Can't be longer than 255 characters (3) Host name segments cannot exceed 63
-   * characters (4) Must end and start with alphanumeric characters
+   * Prompt the user for a valid host name, which follows the below parameters:
```
```suggestion
   * Prompts the user for a valid hostname, which follows the below parameters:
```

rooneyshuman replied:
```suggestion
   * Prompts the user for a valid hostname, which follows the below parameters:
```

carissaallen commented on src/main/java/edu/pdx/cs/sftp/User.java:
```diff
@@ -44,93 +51,97 @@
+   * (4) Must end and start with alphanumeric characters
    *
-   * @return The host name input
+   * @return host name input.
```
```suggestion
   * @return hostname input.
```
I legitimately Googled this (ha!), but I think it should be `hostname` instead of `host name` (see: https://english.stackexchange.com/questions/100858/hostname-or-host-name).

rooneyshuman replied:
```suggestion
   * @return hostname input.
```

carissaallen commented on src/main/java/edu/pdx/cs/sftp/User.java:
```diff
@@ -44,93 +51,97 @@
-      System.out.println(
-          "Valid host names are no longer than 255 alphanumeric characters and \n"
-              + "dashes. Each segment of the host name cannot be longer than 63 characters.");
+      err.println("The host name is invalid.\n"
```
```suggestion
      err.println("The hostname is invalid.\n"
```

carissaallen commented on src/main/java/edu/pdx/cs/sftp/User.java:
```diff
@@ -44,93 +51,97 @@
-          "Valid host names are no longer than 255 alphanumeric characters and \n"
-              + "dashes. Each segment of the host name cannot be longer than 63 characters.");
+      err.println("The host name is invalid.\n"
+                + "Valid host names are no longer than 255 alphanumeric characters and dashes.\n"
```
```suggestion
                + "Valid hostnames are no longer than 255 alphanumeric characters and dashes.\n"
```

carissaallen commented on src/main/java/edu/pdx/cs/sftp/User.java:
```diff
@@ -44,93 +51,97 @@
-              + "dashes. Each segment of the host name cannot be longer than 63 characters.");
+      err.println("The host name is invalid.\n"
+                + "Valid host names are no longer than 255 alphanumeric characters and dashes.\n"
+                + "Each segment of the host name cannot be longer than 63 characters.");
```
```suggestion
                + "Each segment of the hostname cannot be longer than 63 characters.");
```

carissaallen commented on src/main/java/edu/pdx/cs/sftp/User.java:
```diff
@@ -44,93 +51,97 @@
 
   /**
-   * Checks if the string is a valid host name
+   * Verify the host name is alphanumeric, shorter than 255 characters, with name segments not
```
```suggestion
   * Verifies the host name is alphanumeric, shorter than 255 characters, with name segments not
```

rooneyshuman replied:
```suggestion
   * Verifies the hostname is alphanumeric, shorter than 255 characters, with name segments not
```

carissaallen commented on src/main/java/edu/pdx/cs/sftp/User.java:
```diff
@@ -44,93 +51,97 @@
-  boolean verifyHostName(String toVerify) {
-    if (toVerify.length() > 255) return false;
-    String userNamePattern =
+  boolean verifyHostName(String hostnameToVerify) {
```
I think `verifyHostName` should be changed to `verifyHostname` -- the original author used "hostname" and "host name" inconsistently in this class, which is a little distracting. One form should be chosen and used consistently throughout. If you make this change, make sure to change it in _every place_ that `verifyHostName` is called. If you right click on the method name in IntelliJ, click Refactor, and then Rename, it should help you change it in every place. You can also view all the usages of a particular method by placing your cursor on it and hitting cmd+click.

carissaallen commented on src/main/java/edu/pdx/cs/sftp/User.java:
```diff
@@ -4,26 +4,33 @@
+  private Scanner scanner = new Scanner(in);
 
-  /** Class constructor initializing all fields to null */
+  /** Class constructor initializing all fields to null. */
```
Thanks for making punctuation consistent throughout! 👍

rooneyshuman commented on src/main/java/edu/pdx/cs/sftp/User.java:
```diff
@@ -44,93 +51,97 @@
-   * @param password String to be assessed
-   * @return true if the password isn't empty and does not include spaces Otherwise, return false
+   * @param passwordToVerify is the string to be assessed.
+   * @return <code>true</code> if the string is a valid password <code>false</code> otherwise.
```
```suggestion
   * @return <code>true</code> if the password entered is valid; <code>false</code> otherwise.
```

rooneyshuman commented on src/main/java/edu/pdx/cs/sftp/User.java:
```diff
@@ -44,93 +51,97 @@
-   * @param toVerify a string to be checked against a regular expression
-   * @return <code>true</code> if the string is a valid username <code>false</code> otherwise
+   * @param usernameToVerify is the string to be checked against a regular expression.
+   * @return <code>true</code> if the string is a valid username <code>false</code> otherwise.
```
```suggestion
   * @return <code>true</code> if the username entered is valid; <code>false</code> otherwise.
```

carissaallen commented on src/main/java/edu/pdx/cs/sftp/User.java:
```diff
@@ -44,93 +51,97 @@
+   * (1) Must be alphanumeric
+   * (2) Can't be longer than 255 characters
+   * (3) Host name segments cannot exceed 63 characters
+   * (4) Must end and start with alphanumeric characters
```
```suggestion
   * (4) Must start and end with alphanumeric characters
```

rooneyshuman replied:
You have the qualities of a valid hostname listed here numerically, but in `verifyHostname` its all in one sentence: 

> Verify the host name is alphanumeric, shorter than 255 characters, with name segments not exceeding 63 characters, and starts and ends with alphanumeric characters.

I would stick to one of these. Personally, I think the comma-separated list is clearer.

carissaallen commented on src/main/java/edu/pdx/cs/sftp/User.java:
```diff
@@ -44,93 +51,97 @@
+   * Password must not be empty or include spaces.
    *
-   * @return The password input
+   * @return password input.
```
Returns `password input` is not as descriptive as it could be. I would suggest something like: `validated password` or `validated password entered by the user`.

rooneyshuman commented on src/main/java/edu/pdx/cs/sftp/User.java:
```diff
@@ -44,93 +51,97 @@
    *
-   * @param toVerify a string to be checked against a regular expression
-   * @return <code>true</code> if the string is a valid host name <code>false</code> otherwise
+   * @param hostnameToVerify is a string to be checked against a regular expression.
```
```suggestion
   * @param hostnameToVerify input to be checked against a regular expression.
```

rooneyshuman commented on src/main/java/edu/pdx/cs/sftp/User.java:
```diff
@@ -44,93 +51,97 @@
-   * @param toVerify a string to be checked against a regular expression
-   * @return <code>true</code> if the string is a valid host name <code>false</code> otherwise
+   * @param hostnameToVerify is a string to be checked against a regular expression.
+   * @return <code>true</code> if the string is a valid host name <code>false</code> otherwise.
```
```suggestion
   * @return <code>true</code> if the hostname entered is valid; <code>false</code> otherwise.
```

carissaallen reviewed (changes requested):
I made a few small suggestions. Otherwise, the improvements look good! 👍 

One thing I think might be a good area of focus is the verification performed on the username, password, and hostname. They do not seem to work as I would expect them to (at least for me) based on the implementation here. So definitely make sure to run the program and see how it behaves on incorrect inputs. You don't necessarily need to refactor anything, but I would definitely discuss it in your analysis if you experience unexpected behavior from the program.

rooneyshuman reviewed (changes requested):
Some changes to the method descriptors in order for the wording to match the established styleguide. Variable name changes are *much* clearer.

bahamieh committed:
Incorporate team reviews from previous pull request

Takes review change suggestions from previous pull request and implements them.

rooneyshuman reviewed (approved):
LGTM!
